### PR TITLE
Adding support for per call credentials in core.

### PR DESCRIPTION
**Pull request by jboeuf**

### CODE REVIEW

**Comment by jboeuf** [comment]
@yang-g: if you could do the C++ version of this call in ClientContext.h, that would be awesome. Let me know if  you have any questions.

**Comment by jboeuf** [comment]
`./bins/opt/end2end_test` works on my machine. Not sure what's up with python and csharp. Will have a look tomorrow...

**Comment by nicolasnoble** [comment]
#1499 is a fix for C#.

**Comment by ctiller** [comment]
Merge & repush?

**Comment by jboeuf** [comment]
Merged. Building now...

**Comment by jboeuf** [comment]
Yay, we're green!

File: build.json [hunk 1]
```diff
@@ -405,6 +405,7 @@
         "src/core/httpcli/parser.h",
         "src/core/security/auth.h",
         "src/core/security/base64.h",
+        "src/core/security/context.h",
```
Comment (ctiller) [inline build.json]: bikeshed: can we choose a different name than context.{h,c} - visual studio only allows a unique filename to be used once per project, and I feel like context is one of those names that we'd want to be precious about

Comment (ctiller) [inline build.json]: happy with a renaming pull later

File: test/core/end2end/gen_build_json.py [hunk 2]
```diff
@@ -46,34 +46,40 @@
     'chttp2_socket_pair_one_byte_at_a_time': False,
 }
 
-# maps tests names to whether they run fine or not (aka, not flaky)
+class TestOptions(object):
```
Comment (ctiller) [inline test/core/end2end/gen_build_json.py]: TestOptions = collections.namedtuple('TestOptions', ['flaky', 'secure'])

**Comment by ctiller** [comment]
LGTM - Will merge when I can.

**Comment by jboeuf** [comment]
Addressed comments.

**Pull request merged**